pygpu/gpuarray.pyx

```diff
@@ -1465,6 +1465,26 @@ cdef class GpuArray:
         else:
             raise IndexError, "cannot index with: %s" % (key,)
 
+    def write(self, np.ndarray src not None):
+        cdef size_t npsz = np.PyArray_NBYTES(src)
+        cdef size_t sz = gpuarray_get_elsize(self.ga.typecode)
+        cdef unsigned i
+        for i in range(self.ga.nd):
+            sz *= self.ga.dimensions[i]
+        if sz != npsz:
+            raise ValueError, "GpuArray and Numpy array do not have the same size in bytes"
+        array_write(self, np.PyArray_DATA(src), sz)
+
+    def read(self, np.ndarray dst not None):
+        cdef size_t npsz = np.PyArray_NBYTES(dst)
+        cdef size_t sz = gpuarray_get_elsize(self.ga.typecode)
+        cdef unsigned i
+        for i in range(self.ga.nd):
+            sz *= self.ga.dimensions[i]
+        if sz != npsz:
+            raise ValueError, "GpuArray and Numpy array do not have the same size in bytes"
+        array_read(np.PyArray_DATA(dst), sz, self)
```

Comment:
You might want to check that the array is writeable `dst->flags & NPY_ARRAY_BEHAVED) == NPY_ARRAY_BEHAVED`.

`array_read()` will fail if the source is not contiguous (and will scribble everywhere if the destination is not contiguous).  So you might want to check for that.

Also the check done here for compatibility is very basic.  Do you think it could be extended a bit to cover things like compatible shape?